core/src/type_params.rs

```diff
@@ -0,0 +1,307 @@
+uses_type_params!(syn::TypeBareFn, inputs, output);
+uses_type_params!(syn::ParenthesizedGenericArguments, inputs, output);
+uses_type_params!(syn::TraitBound, path);
+uses_type_params!(syn::TypeTraitObject, bounds);
+uses_type_params!(syn::TypeImplTrait, bounds);
+uses_type_params!(syn::Binding, ty);
+
+impl UsesTypeParams for syn::DataEnum {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        self.variants.collect_type_params(type_set)
+    }
+}
+
+impl UsesTypeParams for syn::Fields {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        self.collect_type_params(type_set)
+    }
+}
+
+/// Check if an Ident exactly matches one of the sought-after type parameters.
+impl UsesTypeParams for Ident {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        type_set.iter().filter(|v| *v == self).collect()
+    }
+}
+
+impl UsesTypeParams for syn::ReturnType {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        if let syn::ReturnType::Type(_, ref ty) = *self {
+            ty.uses_type_params(type_set)
+        } else {
+            Default::default()
+        }
+    }
+}
+
+impl UsesTypeParams for Type {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        match *self {
+            Type::Slice(ref v) => v.elem.uses_type_params(type_set),
+            Type::Array(ref v) => v.elem.uses_type_params(type_set),
+            Type::Ptr(ref v) => v.elem.uses_type_params(type_set),
+            Type::Reference(ref v) => v.elem.uses_type_params(type_set),
+            Type::BareFn(ref v) => v.uses_type_params(type_set),
+            Type::Tuple(ref v) => v.uses_type_params(type_set),
+            Type::Path(ref v) => v.uses_type_params(type_set),
+            Type::Paren(ref v) => v.elem.uses_type_params(type_set),
+            Type::Group(ref v) => v.elem.uses_type_params(type_set),
+            Type::TraitObject(ref v) => v.uses_type_params(type_set),
+            Type::ImplTrait(ref v) => v.uses_type_params(type_set),
+            Type::Macro(_) | Type::Verbatim(_) | Type::Infer(_) | Type::Never(_) => {
+                Default::default()
+            }
+        }
+    }
+}
+
+uses_type_params!(syn::TypeTuple, elems);
+
+impl UsesTypeParams for syn::Path {
+    fn uses_type_params<'a>(&self, type_set: &'a IdentSet) -> IdentRefSet<'a> {
+        // Not sure if this is even possible, but a path with no segments definitely
+        // can't use type parameters.
+        // We don't use `is_empty` here because that also considers punctuation.
+        if self.segments.len() == 0 {
+            return Default::default();
+        }
+
+        // A path segment ident can only match if it is not global and it is the first segment
+        // in the path.
+        let ident_hits = if !self.global() {
+            self.segments[0].ident.uses_type_params(type_set)
```

Comment:
@dtolnay is this logic correct, or should I only consider the path's `ident` if the length is exactly 1?

Reply:
If the intention is to use this for trait bounds then this logic is not what you want. For example the struct in your `qself_vec` test case:

```rust
#[derive(SomeTrait)]
struct Foo<T>(<Vec<T> as a::b::Trait>::AssociatedItem);
```

The unit test expects `T` used, but for the purpose of bounds we probably would not want a `T: SomeTrait` bound in this code.

Reply:
Would we never want this field to be "using" the type parameter? I haven't used the syntax enough to have a good sense of when this would be used and if you'd ever need bounds (or type mapping, or something else).

Reply:
FWIW, fitzgen/state_machine_future does track `qself` for exactly the reason I thought. I'm trying to come up with an elegant way of making this work for both; I could try rewriting it as a visitor, but I'm hoping it doesn't come to that.

Another option that comes to mind would be some sort of newtype or tuple struct that sits "around" the type being queried and says "I'm searching for uses to make a new struct/enum". For `TypePath` specifically, it'd intervene and say "consider qself", while in other cases it'd be a passthrough.

A third option would be to add a "purpose" enum and then have that be an extra argument to the `uses_type_params` method. It'd make all uses slightly less ergonomic, but it'd probably be the simplest to implement.

Reply:
Thought about this some more, and I’m going to add an `options` parameter that includes a `purpose` field. That keeps the system extensible and composable, and won’t have much impact on callers.